### PR TITLE
Add errors for passing remote records to extern procs

### LINKED ISSUE #11539: Generate error messages rather than passing remote pointers to externs
### Summary of Problem

At present, the Chapel compiler doesn't protect against passing remote pointers to records off to external routines.  If users do this and try to access the record's fields within their external C routine, they'll likely access garbage memory / values.  This was the root cause of #11288.

The compiler does intentionally narrow such pointers before passing them off to external routines.  In this issue, I'm proposing that it should also do a locality check when `--checks` is on (the default) and print an error if the wide reference isn't local before passing it externally.

### Steps to Reproduce

**Associated Future Test(s):**

[`test/extern/multilocale/externRemoteRecord1.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/extern/multilocale/externRemoteRecord1.chpl
          ) #11538

(as well as `...2.chpl` and `...3.chpl` in that same directory).

### Configuration Information

- Output of `chpl --version`: `chpl version 1.19.0 pre-release (38fb2da)`

### PR DESCRIPTION
**Pull request by bradcray**
This adds an execution time check (when `--local-checks` is thrown, which is the default) for wide references passed to external routines to make sure that they are local to the given locale before stripping out the locale field and passing on the raw address.  The lack of such checks can cause segfaults or unpredictable results since the external routine may access garbage memory, and was the cause of problems for a user in issue #11288.

While here, I refactored the PRIM_LOCAL_CHECK expression to take the error message as its second argument (comments in the code suggested that it was previously the fourth argument, but I saw no indication of this) and to move the default error message which has been used for local block checks in the past into the place where those expressions were generated.

Turning on this check resulted in an error for the case of deinitializing atomic variables from a remote locale.  This hasn't been a problem in practice because the extern atomic_destroy() routine used to deinitialize atomics is a no-op for cases we care most about (and hasn't been tested much on multi-locale executions for ones we care less about like CHPL_ATOMICS=locks).  With Elliot's help diagnosing, I added an on-clause to these cases as well which should make CHPL_ATOMICS=locks work better in multi-locale settings.

While here, I noticed that we still had a lingering error message that was specific to extern classes, but we've removed extern classes, and no remaining tests were exercising this error message, so I removed it as well, simplifying the code slightly.

Working on this, I learned that exported functions also expect their arguments to all be local, preventing them from being called with wide references from within Chapel as well.  This surprised me and I opened issue #11564 to track the issue.  I added a similar execution-time error for this case, though I worded it as though it was a bug since that's how I viewed it and didn't expect disagreement on this issue.  If there is, I could neutralize the wording of the error message if desired.

Resolves #11539

### CODE REVIEW

File: compiler/codegen/expr.cpp [hunk 1]
```diff
@@ -4384,20 +4384,13 @@ DEFINE_PRIM(PRIM_LOCAL_CHECK) {
     // arguments are (wide ptr, line, function/file, error string)
     GenRet lhs = call->get(1);
     Symbol* lhsType = lhs.chplType->symbol;
+    bool hasErrorMsg = (call->numActuals() > 3);
+    const char* error = (hasErrorMsg ?
+                         toVarSymbol(toSymExpr(call->get(2))->symbol())->immediate->v_string :
+                         "cannot access remote data in local block");
```
Comment (mppf) [inline compiler/codegen/expr.cpp]: "cannot access remote data in local block" is just one of many uses of PRIM_LOCAL_CHECK, rather than being inherent or fundamentally the right default error. Partly for this reason, and partly because I think it'll make these changes less confusing, could you update the code in insertWideReferences that adds PRIM_LOCAL_CHECK to always add the error message argument? IWR should be able to construct the string literal for the default error message once and use it in the existing call sites. Then, in this region of code, you'll always have the error as get(2) and the line and file after that. You'll need to also update the comment on line 4384.
Thanks.

File: compiler/passes/insertWideReferences.cpp [hunk 2]
```diff
@@ -1489,6 +1491,11 @@ static void narrowWideClassesThroughCalls()
             stmt->insertBefore(new CallExpr(PRIM_MOVE, var, sym->copy()));
           }
           else if (narrowType.isRef() || narrowType.type()->symbol->hasFlag(FLAG_DATA_CLASS)) {
+            // Insert a local check because we cannot pass narrow references to
+            // remote data to external routines
+            if (!fNoLocalChecks)
+              stmt->insertBefore(new CallExpr(PRIM_LOCAL_CHECK, sym->copy(), buildCStringLiteral(astr("references to remote data cannot be passed to external routines like '", fn->name, "'"))));
```
Comment (mppf) [inline compiler/passes/insertWideReferences.cpp]: Technically, there might be non-extern functions with FLAG_LOCAL_ARGS, so it might be useful to make different error messages depending on whether the function is marked extern. In particular `export` functions also have this flag. You could also reword it to refer to "functions accepting only local arguments".

**Comment by mppf** [comment]
I think you meant to put "Resolves #11288" in the PR message. The current last line points to the other PR adding futures.

**Comment by mppf** [comment]
Not that it matters since that issue is already closed.

**Comment by mppf** [comment]
Ah, you must have meant "Resolves #11539".

File: compiler/passes/insertWideReferences.cpp (hunk 2 above)
Comment (bradcray) [reply compiler/passes/insertWideReferences.cpp]: Oh, thanks for pointing that out.  Note that it actually should be fine for Chapel to pass wide things to exported functions—the requirement/expectation that the incoming arguments are narrow only applies to calls that are actually coming from outside of Chapel.  I'd guess that this probably means we have problems today if we called both using a wide reference from within Chapel and a narrow one from outside of Chapel, but in any case, you've pointed out that my insertion of the check is incorrect for exported functions.

Comment (bradcray) [reply compiler/passes/insertWideReferences.cpp]: I've opened PR #11562 to add a future showing that I don't think export functions should assume all their arguments are local for calls that originate from within Chapel.

File: compiler/codegen/expr.cpp (hunk 1 above)
Comment (bradcray) [reply compiler/codegen/expr.cpp]: Other than the new use I added, "cannot access remote data in local block" is the only other use that we currently have, but I agree that your proposal is more principled and prefer it to what I did.

**Comment by bradcray** [comment]
This is currently failing for the following tests, complaining of a remote reference being passed into atomic_destroy():

library/packages/Collection/CollectionBulk.chpl
library/packages/Collection/CollectionCounter.chpl
library/packages/Collection/CollectionNQueens.chpl
library/packages/Collection/DequeOrdering.chpl
library/packages/Collection/DequeParity.chpl
library/packages/ZMQ/multilocale/hello-multilocale.chpl
optimizations/remoteValueForwarding/alwaysRVFpragma.chpl
release/examples/benchmarks/ssca2/SSCA2_main.chpl
studies/ssca2/test-rmatalt/nondet.chpl

I think talking through this with @ronawho tomorrow would likely help.

**Comment by bradcray** [comment]
@mppf: I've updated the code and PR message and believe this is ready for re-review.  Be sure to check and see whether you'd like a different wording for the remote reference to export function case given that you seem to have a different perspective on what should / should not be legal there.

(I should also be able to now add a .bad file for the export future I put together this weekend, but haven't done so yet).

File: compiler/passes/insertWideReferences.cpp [hunk 3]
```diff
@@ -1482,13 +1484,23 @@ static void narrowWideClassesThroughCalls()
...
           }
           else if (narrowType.isRef() || narrowType.type()->symbol->hasFlag(FLAG_DATA_CLASS)) {
+
+            // Insert a local check because we cannot pass narrow references to
+            // remote data to external routines
+            if (!fNoLocalChecks) {
+              if (fn->hasFlag(FLAG_EXTERN))
+                stmt->insertBefore(new CallExpr(PRIM_LOCAL_CHECK, sym->copy(), buildCStringLiteral(astr("references to remote data cannot be passed to external routines like '", fn->name, "'"))));
+              else if (fn->hasFlag(FLAG_EXPORT))
+                stmt->insertBefore(new CallExpr(PRIM_LOCAL_CHECK, sym->copy(), buildCStringLiteral(astr("references to remote data cannot currently be passed to exported routines like '", fn->name, "' (though this is a bug)"))));
```
Comment (mppf) [inline compiler/passes/insertWideReferences.cpp]: I'd definitely leave the "Though this is a bug" until we're all convinced of that. "Cannot currently" is sufficient I think.

Comment (mppf) [reply compiler/passes/insertWideReferences.cpp]: You could say "(Though this may change as the implementation improves)" or something along those lines.

Comment (bradcray) [reply compiler/passes/insertWideReferences.cpp]: With fresh eyes today, I agree.

**Comment by ronawho** [comment]
While doing some other testing, I noticed this breaks network atomics. 

There we pass the owning locale and address (by passing the _v var by ref/const ref) to an extern function and then the runtime takes care of things. I think this is a valid use for passing a remote reference to an extern proc, but maybe there's a different way we could calculate the address of _v. Either that or we need a pragma or something to mark this as safe. Maybe something like https://github.com/chapel-lang/chapel/compare/master...ronawho:pragma-remote-extern-var-ok ?

**Comment by ronawho** [comment]
Instead of the pragma approach, we decided to go with https://github.com/chapel-lang/chapel/pull/11579, which uses `__primitive("_wide_get_addr", _v)` to explicitly calculate the address of `_v` instead of passing `_v` by `ref` and relying on the compiler to do the equivalent thing.
